core/src/main/java/com/linecorp/armeria/server/http/auth/HttpAuthService.java

```diff
@@ -91,10 +92,20 @@ protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req) thr
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        if (authorize(req.headers())) {
-            return onSuccess(ctx, req);
-        } else {
-            return onFailure(ctx, req);
-        }
+        DeferredHttpResponse res = new DeferredHttpResponse();
+
+        authorize(req, ctx).whenComplete((result, t) -> {
+            try {
+                if (t != null || !result) {
+                    res.delegate(onFailure(ctx, req));
+                } else {
+                    res.delegate(onSuccess(ctx, req));
+                }
+            } catch (Exception e) {
+                res.close(e);
```

Comment:
This is for exceptions thrown during onSuccess / onFailure (exceptions thrown during authorization are handled by ```t != null```). Since callbacks can't throw checked exceptions, I need to squash here.